File: music-haskell/compiler.hs
import Data.Char

data Expression =
    Seq { left :: Expression, right :: Expression }
  | Note { pitch :: String, pitchNumber :: Integer, duration :: Integer, start :: Integer }
  | Par { left :: Expression, right :: Expression }
  | Rest { duration :: Integer }
  | Repeat { section :: Expression, count :: Integer }
  deriving (Show)

defaultNote = Note { pitch = "a0", pitchNumber = 0, duration = 0, start = 0 }

endTime :: Expression -> Integer -> Integer
endTime (Note { duration = duration }) start = duration + start
endTime (Seq { left = left, right = right }) start = endTime right (endTime left start)
endTime (Par { left = left, right = right }) start = max (endTime left start) (endTime right start)
endTime (Rest { duration = duration }) start = duration + start
endTime (Repeat { count = count, section = section }) start = start + count * endTime section 0

convertPitch :: String -> Integer
convertPitch (letter:number) = 12 + 12 * octave + letterPitch letter
  where
    octave = read number :: Integer
    letterPitch letter = [9,11,0,2,4,5,7] !! (ord letter - ord 'a')

compile :: Expression -> [Expression]
compile expression = compile' expression 0
  where
    compile' :: Expression -> Integer -> [Expression]
    compile' (Seq { left = left, right = right }) start = compile' left start ++ compile' right (endTime left start)
    compile' (Par { left = left, right = right }) start = compile' left start ++ compile' right start
    compile' (Note { pitch = pitch, duration = duration }) start = [Note { pitch = pitch, pitchNumber = convertPitch pitch, duration = duration, start = start }]
    compile' (Rest r) start = []
    compile' (Repeat { count = count, section = section }) start = foldl (++) [] $ map rep [0..count - 1]
      where
        rep i = compile' section (start + i * endTime section 0)

playNote :: [Expression] -> [Expression]
playNote expression = expression

playMusic :: Expression -> [Expression]
playMusic = playNote . compile

toCompileSeq = Seq {
    left = Seq {
        left = defaultNote { pitch = "a4", duration = 250 },
        right = defaultNote { pitch = "b4", duration = 250 }
      },
    right = Seq {
        left = Seq {
          left = defaultNote { pitch = "c4", duration = 500 },
          right = Rest { duration = 250 }
        },
        right = Repeat {
          count = 3,
          section = defaultNote { pitch = "d4", duration = 500 }
        }
      }
  }

toCompilePar = Seq {
    left = Par {
        left = defaultNote { pitch = "c3", duration = 250 },
        right = defaultNote { pitch = "g4", duration = 500 }
      },
    right = Par {
        left = defaultNote { pitch = "d3", duration = 500 },
        right = defaultNote { pitch = "f4", duration = 250 }
      }
  }

main = do
  putStrLn $ show toCompileSeq
  putStrLn $ show $ compile toCompileSeq
  putStrLn $ show toCompilePar
  putStrLn $ show $ compile toCompilePar

File: music/compiler.js
var endTime = function(expr, time) {
  if(expr.tag === 'note') {
    return expr.duration + time;
  } else if(expr.tag === 'seq') {
    return endTime(expr.right, endTime(expr.left, time));
  } else if(expr.tag === 'par') {
    return Math.max(endTime(expr.left, time), endTime(expr.right, time));
  }
}
var _compile = function(expr, start) {
  if(expr.tag === 'seq') {
    var left = _compile(expr.left, start);
    var rightStart = endTime(expr.left, start);
    var right = _compile(expr.right, rightStart);
    return left.concat(right);
  } else if(expr.tag === 'par') {
    var left = _compile(expr.left, start);
    var right = _compile(expr.right, start);
    return left.concat(right);
  } else if(expr.tag === 'note') {
    return [ {
      tag: 'note',
      start: start,
      pitch: expr.pitch,
      duration: expr.duration
    } ];
  }
};

var compile = function(expr) {
  return _compile(expr, 0);
}

var playNote = function() {};

var playMusic = function(expr) {
  playNote(compile(expr));
}

var melody_music =
    { tag: 'seq',
      left:
       { tag: 'seq',
         left: { tag: 'note', pitch: 'a4', duration: 250 },
         right: { tag: 'note', pitch: 'b4', duration: 250 } },
      right:
       { tag: 'seq',
         left: { tag: 'note', pitch: 'c4', duration: 500 },
         right: { tag: 'note', pitch: 'd4', duration: 500 } } };

console.log(melody_music);
console.log(compile(melody_music));

var melody_parallel =
    { tag: 'seq',
      left:
       { tag: 'par',
         left: { tag: 'note', pitch: 'c3', duration: 250 },
         right: { tag: 'note', pitch: 'g4', duration: 500 } },
      right:
       { tag: 'par',
         left: { tag: 'note', pitch: 'd3', duration: 500 },
         right: { tag: 'note', pitch: 'f4', duration: 250 } } };

console.log(melody_parallel);
console.log(compile(melody_parallel));
